### ISSUE TITLE
Linear Interpolation Additions: unlerp / relerp

### ISSUE THREAD

**Issue by dicksonlaw583** [Alphish/gm-community-toolbox#8]
These two functions occur so frequently in situations involving linear interpolation that I wished they were built-in. Especially `relerp`, which I cite on the GMC all the time for things like health and scroll bars, room-to-screen coordinate conversions, etc.

```gml
///@func unlerp(a, b, v)
///@param {Real} a The first value.
///@param {Real} b The second value.
///@param {Real} v The interpolated value.
///@return {Real}
///@desc Return the interpolation coefficient t such that lerp(a, b, t) equals v.
function unlerp(a, b, v) {
    return (v-a)/(b-a);
}
```

```gml
///@func relerp(x0, x1, v, y0, y1)
///@param {Real} x0 The old range's first value.
///@param {Real} x1 The old range's second value.
///@param {Real} v The value to re-interpolate.
///@param {Real} y0 The new range's first value.
///@param {Real} y1 The new range's second value.
///@return {Real}
///@desc Given v in the old range (x0, x1), return its equivalent value in the new range (y0, y1).
function relerp(x0, x1, v, y0, y1) {
    return y0+(y1-y0)*(v-x0)/(x1-x0);
}
```

**Comment by Alphish** [comment]
Thanks, definitely planning to add functions like these (I doubt anyone would complain)!
I considered names like `lerp_invert` and `lerp_remap`, but `unlerp` and `relerp` are both shorter and still hard to mistake for anything else! ^^

I'm considering more descriptive argument names, like `unlerp(from,to,value)` and `relerp(oldfrom,oldto,oldvalue,newfrom,newto)`

Note: when it comes to argument names specifically, I prefer to keep them all lowercase and thus two words at most (somehow, underscore or uppercase letters look odd to me in GML function previews), but I'm willing to settle on something else. Might draft an issue about coding conventions later on...

Would be good to decide on some error handling for when either range is zero (i.e. b = a in unlerp and x0 = x1 or y0 = y1 in relerp). Do we throw an exception? And if we do, do we throw exceptions as a plain string (e.g. `throw "The relerp old range has 0 width.";`), or something more elaborate?

**Comment by dicksonlaw583** [comment]
Here are the more descriptive versions should you decide to implement them:

```gml
///@func unlerp(from, to, value)
///@param {Real} from The first value.
///@param {Real} to The second value.
///@param {Real} value The interpolated value.
///@return {Real}
///@desc Return the interpolation coefficient t such that lerp(from, to, t) equals the given value.
function unlerp(from, to, value) {
    if (from == to) {
        throw "unlerp() range has zero width.";
    }
    return (value-from)/(to-from);
}
```

```gml
///@func relerp(oldfrom,oldto,oldvalue,newfrom,newto)
///@param {Real} oldfrom The old range's first value.
///@param {Real} oldto The old range's second value.
///@param {Real} oldvalue The value to re-interpolate.
///@param {Real} newfrom The new range's first value.
///@param {Real} newto The new range's second value.
///@return {Real}
///@desc Given oldvalue in the old range (oldfrom, oldto), return its equivalent value in the new range (newfrom, newto).
function relerp(oldfrom,oldto,oldvalue,newfrom,newto) {
    if (oldto == oldfrom) {
        throw "relerp() old range has zero width.";
    }
    return newfrom+(newto-newfrom)*(oldvalue-oldfrom)/(oldto-oldfrom);
}
```

I think we should save the discussion on error handling for #11, as this isn't the only place where invalid input may arise. But do note that y0 = y1 is NOT an error condition in `relerp`, it is a degenerate but still valid use case.

**Comment by gnysek** [comment]
There was similar discussion after naming convention of `unlerp` on GMC last year and `inverse_lerp` was among winners.

https://forum.gamemaker.io/index.php?threads/code-bank-share-useful-code.60575/page-3#post-598508
https://forum.gamemaker.io/index.php?threads/code-bank-share-useful-code.60575/page-3#post-598578

**Comment by Alphish** [comment]
`inverse_lerp` didn't really *win*. JuJu just mentioned something like "invers lerp" and thought it was a made up thing, and I pointed out "inverse lerp" is actually what people call this function (but it referred to how people call the concept "inverse lerp" rather than the function identifier `inverse_lerp`).

Personally, I prefer `unlerp` and `relerp` as they are briefer, and yet still descriptive enough to people who are familiar with two `lerp` counterparts.

**Comment by Alphish** [comment]
Alright, there doesn't seem to be much discussion here, and aside from my `lerp_invert` and `lerp_remap` ideas - which I don't prefer at all - and `inverse_lerp` that was more of a misunderstanding of a thread discussion, there aren't other particular name ideas.

As for error handling, I suggest we settle for throwing simple strings for now. If we decide we want to throw more advanced errors, we can always change it later.

Thus, let's get to basic reactions voting:
- 🎉 `unlerp` and `relerp`
- 😕 something else altogether

**Comment by Alphish** [comment]
Since no-one disagrees with the `unlerp` and `relerp` naming, we'll use these for the implementation. As a reminder, the functions should check for the `unlerp`/`relerp` source range being of 0 width, and throw an exception accordingly (the destination range should be fine).

The tests should check different unlerp/relerp scenarios, including 0-width source range.

The demonstration would likely involve some sliders of the source/target ranges and also the current value.

**Comment by Mercerenies** [comment]
For what it's worth, Godot uses [`inverse_lerp`](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#class-globalscope-method-inverse-lerp) and [`remap`](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#class-globalscope-method-remap). Though I'm *definitely* not a fan of `remap`, so `unlerp` and `relerp` seem pretty good to me.

**Comment by Alphish** [comment]
The implementation and tests have been already merged, now only the demo remains.

**Comment by Alphish** [comment]
The demo has been merged in Pull Request #31 as well!
